src/lib/Names.js

```diff
@@ -0,0 +1,211 @@
+    name = name.replace(CLEAN_REGEX, '/');
+
+    if (name.endsWith('/')) {
+      return name.substr(0, -1);
+    }
+    // else
+    return name;
+  }
+
+  append(left, right) {
+    return this.clean(left + '/' + right);
+  }
+
+  remap(name) {
+    return this.resolve(name, true);
+  }
+
+  /**
+   * @param [namespace] {string} namespace to resolve name to. If not provided, node's namespace will be used
+   * @param name {string} name to resolve
+   * @param [remap] {bool} flag indicating if we should also attempt to remap the name
+   */
+  resolve(...args) {
+    let [namespace, name, remap] = this._parseResolveArgs(args);
+
+    this.validate(name, true);
+
+    if (name.length === 0) {
+      if (namespace.length === 0) {
+        return '/';
+      }
+      else if (namespace[0] === '/') {
+        return namespace
+      }
+      // else
+      return '/' + namespace;
+    }
+
+    if (name.startsWith('~')) {
+      name = name.replace('~', this._namespace + '/');
+    }
+
+    if (!name.startsWith('/')) {
+      name = namespace + '/' + name;
+    }
+
+    name = this.clean(name);
+
+    if (remap) {
+      name = this._remap(name);
+    }
+
+    return name;
+  }
+
+  parentNamespace(name) {
+    this.validate(name, true);
+
+    if (name.length === 0) {
+      return '';
+    }
+    else if (name === '/') {
+      return '/';
+    }
+
+    let p = name.lastIndexOf('/');
+    if (p === name.length - 1) {
+      p = name.lastIndexOf('/', p - 1);
+    }
+
+    if (p < 0) {
+      return '';
+    }
+    else if (p === 0) {
+      return '/';
+    }
+    // else
+    return name.substring(0, p);
+  }
+
+  _remap(name) {
+    return this._remappings[name] || name;
+  }
+
+  _parseResolveArgs(args) {
+    let name, namespace = this._namespace, remap = true;
+    switch (args.length) {
+      case 0:
+        name = '';
+        break;
+      case 1:
+        name = args[0];
+        break;
+      case 2:
+        if (typeof args[1] === 'string') {
+          [namespace, name] = args;
+        }
+        else {
+          [name, remap] = args;
+        }
+        break;
+      default:
+        [namespace, name, remap] = args;
+        break;
+    }
+
+    return [namespace, name, remap];
+  }
+}
+
+module.exports = new Names();
+
+//------------------------------------------------------------------
+// Local Helper functions
+//------------------------------------------------------------------
+
+
+
+function isAlpha(char, code = null) {
+  code = code || char.charCodeAt(0);
+  if (code >= 65 && code <= 90) {
+    return true;
+  }
+  else if (code >= 97 && code <= 122) {
+    return true;
+  }
+  // else
+  return false;
+}
+
+function isAlnum(char) {
+  const code = char.charCodeAt(0);
+  if (isAlpha(char, code)) {
+    return true;
+  }
+  else if (code >= 48 && code <= 57) {
+    return true;
+  }
+  // else
+  return false;
+}
+
+function isValidCharInName(char) {
+  if (isAlnum(char) || char == '/' || char == '_') {
+    return true;
```

Comment:
This code seems overly clear. The minor suggestion is to remove the return and return the condition immediately. i.e

```
return (isAlnum(char) || char == '/' || char == '_')
```

The major suggestion is to use strict comparison: 

```
isAlnum(char) || char === '/' || char === '_')
```